src/sentry_sdk.cpp

```diff
@@ -0,0 +1,139 @@
+#include "sentry_sdk.h"
+
+#include "sentry/contexts.h"
+#include "sentry/disabled_sdk.h"
+#include "sentry/uuid.h"
+#include "sentry_options.h"
+
+#include <godot_cpp/classes/os.hpp>
+#include <godot_cpp/variant/utility_functions.hpp>
+
+#ifdef NATIVE_SDK
+#include "sentry/native/native_sdk.h"
+#endif
+
+using namespace godot;
+using namespace sentry;
+
+SentrySDK *SentrySDK::singleton = nullptr;
+
+VARIANT_ENUM_CAST(Level);
+
+void SentrySDK::capture_message(const String &p_message, Level p_level, const String &p_logger) {
+	internal_sdk->capture_message(p_message, p_level, p_logger);
+}
+
+void SentrySDK::add_breadcrumb(const String &p_message, const String &p_category, Level p_level,
+		const String &p_type, const Dictionary &p_data) {
+	internal_sdk->add_breadcrumb(p_message, p_category, p_level, p_type, p_data);
+}
+
+String SentrySDK::get_last_event_id() const {
+	return internal_sdk->get_last_event_id();
+}
+
+void SentrySDK::set_tag(const String &p_key, const String &p_value) {
+	ERR_FAIL_COND_MSG(p_key.is_empty(), "Sentry: Can't set tag with an empty key.");
+	internal_sdk->set_tag(p_key, p_value);
+}
+
+void SentrySDK::remove_tag(const String &p_key) {
+	ERR_FAIL_COND_MSG(p_key.is_empty(), "Sentry: Can't remove tag with an empty key.");
+	internal_sdk->remove_tag(p_key);
+}
+
+void SentrySDK::set_user(const Ref<SentryUser> &p_user) {
+	ERR_FAIL_NULL_MSG(p_user, "Sentry: Setting user failed - user object is null. Please, use Sentry.remove_user() to clear user info.");
+
+	// Initialize user ID if not supplied.
+	if (p_user->get_id().is_empty()) {
+		// Take user ID from the runtime config or generate a new one if it's empty.
+		String user_id = get_user()->get_id();
+		if (user_id.is_empty()) {
+			user_id = sentry::uuid::make_uuid();
+		}
+		p_user->set_id(user_id);
+	}
+
+	// Save user in a runtime conf-file.
```

Comment:
Why would we do this? Not sure it's the right path. If the game crashes, we shouldn't be reporting the previous user's name on follow up events (other than that one crash itself of course)

Reply:
So the user_id is more like a session_id? Each time generated anew?

Reply:
usually user is set by the developer once a user logs in. For example.

On mobile we have an 'id' that's fixed (the installation id), we use that on release health so we can get a count of users affected.

So if this is that 'installation_id', it makes sense to reuse the same id.

Note that user has also email, and username. That's what I meant by not storing on disk and reloading. The id (installation_id) is probably fine although I'm not srue we do this on desktop (could I have a game where different users login?)

Reply:
Lets figure out what to do here

Reply:
So currently we simply generate a random UUID and persist it for consistency’s sake. By default, that is the only thing that is generated and persisted. But if a developer is [setting the data manually](https://github.com/getsentry/sentry-godot/blob/main/src/sentry_user.h) (like email, username, and even ID) - that data is saved instead. The idea is that the dev pops up a custom form, gets the necessary bits and the user permission, sets it using the public API, and doesn't have to do it again or bother about its persistence. I am not really sure if this is the right approach, and games might have different requirements. But the thing is, if we would rely on dev calling `set_user` every time, it might come too late if an event shows up early in the app's lifecycle.

I was considering having this persistence optional, such as using `persist_user_data` boolean flag in the options. This would simply generate a fresh `user_id` each time the app is started, and not persist it. So it would behave like a session_id with that option off. What do you think? Just for clarity, we don't do IO constantly - the values are loaded from the file on start and cached in memory for later use.